src/foam/u2/search/FilterController.js

```diff
@@ -0,0 +1,524 @@
+      class: 'StringArray',
+      name: 'searchFields',
+      documentation: 'Property names that are currently selected as filters.',
+      factory: function() { return []; },
+      postSet: function(old, nu) {
+        // Check for every filter that has been removed, and every filter that
+        // is freshly added.
+        // This function is responsible for choosing the view for each property.
+        // Eg. drop-downs for Booleans and Enums, before/after for dates, etc.
+        if ( old ) {
+          for ( var i = 0; i < old.length; i++ ) {
+            if ( ! nu || nu.indexOf(old[i]) < 0 ) {
+              this.searchMgr_.remove(old[i]);
+              this.searchViews_[old[i]].remove();
+              delete this.searchViews_[old[i]];
+            }
+          }
+        }
+
+        if ( nu ) {
+          for ( var i = 0; i < nu.length; i++ ) {
+            if ( ! old || old.indexOf(nu[i]) < 0 ) {
+              var split = this.splitName(nu[i]);
+              var prop = this.data.of.getAxiomByName(split.name);
+              var spec = prop.searchView;
+              // TODO(braden): Bring in date support when it's ready.
+              var options = {
+                inline: true,
+                name: nu[i]
+              };
+              if ( prop.tableSeparator ) {
+                options.split = prop.tableSeparator;
+              }
+              this.addGroup(spec, prop, options);
+              this.renderFilter(nu[i]);
+            }
+          }
+        }
+      }
+    },
+    {
+      name: 'searchViews_',
+      factory: function() { return {}; }
+    },
+    {
+      name: 'search',
+      factory: function() {
+        var of = this.data.of;
+        if ( of.id ) of = of.id;
+        return this.TextSearchView.create({
+          of: of,
+          richSearch: true,
+          keywordSearch: true
+        });
+      }
+    },
+    {
+      name: 'filtersE_',
+      factory: function() {
+        var e = this.start().cssClass(this.myCls('filters'));
+        e.end();
+        return e;
+      }
+    }
+  ],
+
+  methods: [
+    function initE() {
+      this.searchMgr_.predicate$.sub(this.onPredicateChange);
+      this.filteredDAO$.sub(this.onPredicateChange);
+      this.onPredicateChange();
+
+      this.cssClass(this.myCls());
+      this.startContext({ data: this });
+      var topPanel = this.start().cssClass(this.myCls('search-panel'))
+          .start().cssClass(this.myCls('adding'));
+      if ( this.allowAddingFilters ) {
+        topPanel.start().cssClass(this.myCls('add-filter'))
+            .start(this.Select, {
+              inline: true,
+              data$: this.filterChoice$,
+              choices: this.filters
+            }).end()
+            .add(this.NEW_FILTER)
+            .end();
+      }
+
+      topPanel.start().cssClass(this.myCls('count'))
+          .start('span')
+              .cssClass(this.myCls('count-text'))
+              .add(this.countString$)
+              .end()
```

Comment:
I think starts and ends should be on the same level. This style is difficult to read.